main/src/library/Graph.flix

```diff
@@ -560,4 +560,153 @@ mod Graph {
+            let Path(ys) = y;
+            if (List.length(xs) <= List.length(ys)) x else y
+    }
+
+    instance MeetLattice[Path] {
+        pub def greatestLowerBound(x: Path, y: Path): Path = ???
+    }
+
+
+    pub def init(y: Int32, x: Int32): Path =
+        Path(y :: x :: Nil)
+
+    pub def cons(z: Int32, p: Path): Path = match p {
+        case Path(xs) => Path(z :: xs)
+    }
+
+    pub def done(p: Path): Bool = match p {
+        case Path(xs) => List.isEmpty(xs)
+    }
+
+    type alias From = Int32
+    type alias To = Int32
+    type alias UpperCapacity = Int32
+    type alias Flow = Int32
+    type alias FlowNetwork = (From, To, UpperCapacity, Flow)
+
+    // TODO: Handle residual network, i.e., reversed edges with negative flow and allow finding paths in that combined with normal network.
+
+    pub def maxFlow(src: { src = From }, dst: { dst = To }, g: m[(From, To, UpperCapacity)]): Flow with Foldable[m] =
+        def fordFulkerson(flowNetwork) = {
+            let path = augmentingPath(src, dst, flowNetwork);
+            if (done(path))
+                getMaxFlow(dst.dst, flowNetwork)
+            else
+                let incr = pathMinCapacity(path, flowNetwork);
+                let updatedNetwork = increaseFlow(path, incr, flowNetwork);
+                fordFulkerson(updatedNetwork)
+        };
+        // Init with 0 flow
+        let flowNetwork = Foldable.toVector(g) |> Vector.map(match (x, y, z) -> (x, y, z, 0));
+        fordFulkerson(flowNetwork)
+
+    def getMaxFlow(dst: To, g: m[FlowNetwork]): Flow with Foldable[m] =
+        let edges = inject g into Edge;
+        let rules = #{
+            Flow(f) :- Edge(_, dst, _, f).
+        };
+        let result = query edges, rules select f from Flow(f);
+        Vector.sum(result)
+
+    def augmentingPath(src: { src = From }, dst: { dst = To }, g: m[FlowNetwork]): Path with Foldable[m] =
+        let edges = inject g into Edge;
+        let rules = #{
+            Reach(x, y; init(y, x)) :- Edge(x, y, u, f), if (u - f) > 0.
+            Reach(x, z; cons(z, p)) :- Reach(x, y; p), Edge(y, z, u, f), if (u - f) > 0.
+        };
+        let s = src.src;
+        let t = dst.dst;
+        let result = query edges, rules select fn from Reach(s, t; fn);
+        Vector.get(0, result)
+
+    def pathMinCapacity(p: Path, g: m[FlowNetwork]): Int32 with Foldable[m] =
+        let Path(xs) = p;
+        let ps = inject xs into Path;
+        let edges = inject g into Edge;
+        let rules = #{
+            PathCapacity(u - f)   :- Edge(x, y, u, f), Path(x), Path(y), if f <= u and f >= 0.
+            NonMinCapacity(u)     :- PathCapacity(u), PathCapacity(v), if v < u.
+            MinCapacity(u)        :- PathCapacity(u), not NonMinCapacity(u).
+        };
+        let result = query ps, edges, rules select u from MinCapacity(u);
+        Vector.get(0, result)
+
+    def increaseFlow(p: Path, incr: Int32, g: m[FlowNetwork]): Vector[FlowNetwork] with Foldable[m] =
```

Comment:
Ok sure, but I think a map on g could also suffice? :)

Reply:
Yeah I agree but checking that the edge is on the given path is easier in datalog

Reply:
But I guess it's just a call to filter first

Reply:
memberOf